nixos/lib/test-driver/test_driver/driver.py

```diff
@@ -101,6 +106,122 @@ def __exit__(self, *_: Any) -> None:
+
+            # delete machines that are no longer part of the test
+            to_del = []
+            for idx, machine in enumerate(self.machines):
+                if machine.name not in names:
+                    if machine.is_up():
+                        self.logger.warning(
+                            f"{machine.name} removed from the test, but it's running so we're not going to shut it down automatically. Call driver.machines[\"{machine.name}\"].stop() and re-run rebuild() to remove."
+                        )
+                    else:
+                        to_del.append(idx)
+            for idx in sorted(to_del, reverse=True):
+                self.logger.info(
+                    f"{self.machines[idx].name} removed from the test. deleting it from the environment"
+                )
+                del globals()[pythonize_name(self.machines[idx].name)]
+                del self.machines[idx]
+
+            # add and change new machines
+            for start_cmd in start_cmds:
+                existing = [
+                    m for m in self.machines if m.name == start_cmd.machine_name
+                ]
+                if len(existing) == 0:
+                    machine = self.create_machine(start_cmd._cmd)
+                    name = pythonize_name(machine.name)
+                    globals()[name] = machine
+                    self.machines.append(machine)
+
+                    self.logger.info(
+                        f'new machine created {start_cmd.machine_name}. start it with driver.machines["{start_cmd.machine_name}"].start()'
+                    )
+                elif len(existing) == 1:
+                    machine = existing[0]
+                    if machine.start_command._cmd == start_cmd._cmd:
+                        self.logger.info(f"{start_cmd.machine_name} unchanged.")
+                    elif not machine.booted:
+                        self.logger.info(f"{start_cmd.machine_name} changed.")
+                        machine.start_command = start_cmd
+                    else:
+                        with self.logger.nested(
+                            f"{start_cmd.machine_name} updated. switching to new configuration"
+                        ):
+                            # TODO: assuming things are at well-known paths isn't great
+                            switch_cmd = (
+                                Path(start_cmd._cmd).parent.parent
+                                / "system"
+                                / "bin"
+                                / "switch-to-configuration"
+                            )
+                            machine.succeed(f"{switch_cmd} test")
+                            machine.start_command = start_cmd
+                else:
+                    self.logger.warning(
+                        f"{start_cmd.machine_name} has multiple instances. This shouldn't be possible, but either way it's now ambiguous which to update."
```

Comment:
Could we log info about the running instances?

Reply:
I'm not sure how this would happen so I'm not sure what would be useful info. If it does happen it's certainly a bug in the testing framework and not a problem with the user's code. Maybe I should say that explicitly.